File: 402-remove-k-digits/402-remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        Stack<Character> st = new Stack<>();
        int size = num.length();
        if(k==size) return "0";
        int counter = 0;
        while(counter<size) {
            while(k>0 && !st.isEmpty() && st.peek()>num.charAt(counter)) {
                st.pop();
                k--;
            }
            st.push(num.charAt(counter));
            counter++;
        }
        //to handle case 11111
        while(k>0) {
            st.pop();
            k--;
        }
        StringBuilder ans = new StringBuilder();
        while(!st.isEmpty()) {
            char c = st.pop();
            ans.append(c);
        }
        ans.reverse();
        //to handle 00012 case ie make it 12 or 000 to 0
        while(ans.length()>1 && ans.charAt(0)=='0') {
            ans.deleteCharAt(0);
        }
        
        return ans.toString();
    }
}

File: 402-remove-k-digits/402-remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) 
    {
         if(k == num.length()) return "0";
        
         stack<char> stk;
         
          int i = 0;
        
          while(i < num.length())
         {
              while(k>0 && !stk.empty() && stk.top()>num[i]) {
                  stk.pop();
                  k--;
              }
              stk.push(num[i]);
              i++;
         }
        //to handle 111 and k=2 to get 1 as res
        while(k>0) 
        {
            stk.pop();
            k--;
        }
            
          string res;
          while(!stk.empty())
          {
              res += stk.top();
              stk.pop();
          }
          
          reverse(res.begin(), res.end());
          //to handle leading zeros 00012 should become 12 or 0000 should become 0
        while(res.length()>1 && res[0]=='0') {
            res.erase(res.begin());
        }
        
           if(res == "") return "0";
           return res;
    }
};